### PR TITLE
DrivAerML: T_max=50 + gc=1.0 compound (stabilized long cycles)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Your T_max sweep (#3045) showed T_max=50 found a val=10.41% pre-divergence basin at ep58 — the gradient explosion (max norm 4.5e8) was the problem, not the basin depth. Adding gc=1.0 should prevent the gradient explosion while preserving the longer exploration cycles of T_max=50.

This is a compound hypothesis: T_max=50 provides more time in each loss basin + gc=1.0 clips the explosive gradient updates at high-LR restart phases. The key finding from #3045 was that T_max=30 is uniquely stable WITHOUT gc — but WITH gc, longer periods may access deeper basins.

## Instructions

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 50 \
  --grad-clip 1.0 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group dm-tmax-gc-compound
```

Note: gc=1.0 was previously tested alone on DM (#3010) — val=4.54% (worse than 3.997%). But that was at T_max=30. The compound with T_max=50 is a new configuration.

If gc=1.0 is stable, also try gc=0.5:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 50 \
  --grad-clip 0.5 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group dm-tmax-gc-compound
```

## Reporting Requirements
- Best val `surface_rel_l2_pct` + epoch for each variant
- Whether either run diverged (and at which epoch)
- Test at best val checkpoint
- W&B run IDs

## Baseline

**DrivAerML champion:** val=**3.997%** @ ep467 (PR #2898, T_max=30, no gc)
- W&B: `bht6h42t`
- Your #3045 T_max=50 without gc: val=10.41% (diverged ep~60)
- Previous gc=1.0 alone at T_max=30: val=4.54% (#3010, gc adds damping)
- External: Transolver-3 3.71% | AB-UPT 3.82%

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT griffith:

## Results

### Variant 1: T_max=50 + gc=1.0

| Metric | Value |
|--------|-------|
| Best val `surface_rel_l2_pct` | **12.56%** @ epoch 42 |
| Test `surface_rel_l2_pct` (best checkpoint) | **13.67%** |
| Diverged at epoch | ~52 (grad norm explosion, reached Infinity by ep80) |
| Total epochs run | 94 (360-min timeout) |
| W&B run | `4ixc5d0e` |
| Peak memory | N/A (not logged in output) |

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 50 \
  --grad-clip 1.0 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group dm-tmax-gc-compound \
  --agent griffith --wandb-name "griffith/dm-tmax50-gc10"
```

### Variant 2: T_max=50 + gc=0.5

| Metric | Value |
|--------|-------|
| Best val `surface_rel_l2_pct` | **7.28%** @ epoch 107 |
| Test `surface_rel_l2_pct` (best checkpoint) | **8.69%** |
| Diverged at epoch | ~121 (grad norm explosion, reached Infinity by ~ep135) |
| Total epochs run | 480 (360-min timeout) |
| W&B run | `0pux5dkr` |
| Peak memory | N/A (not logged in output) |

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 50 \
  --grad-clip 0.5 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group dm-tmax-gc-compound \
  --agent griffith --wandb-name "griffith/dm-tmax50-gc05"
```

### Comparison against baseline

| Config | Best Val | Test (best ckpt) | Diverged? | Stable epochs |
|--------|----------|-------------------|-----------|---------------|
| **Baseline** (T_max=30, no gc) | **3.997%** @ ep467 | — | No | 467+ |
| T_max=50 + gc=1.0 | 12.56% @ ep42 | 13.67% | Yes @ ep52 | ~51 |
| T_max=50 + gc=0.5 | 7.28% @ ep107 | 8.69% | Yes @ ep121 | ~120 |
| T_max=50, no gc (#3045) | 10.41% @ ep58 | — | Yes @ ep~60 | ~59 |
| External: Transolver-3 | — | 3.71% | — | — |
| External: AB-UPT | — | 3.82% | — | — |

### What happened

**The hypothesis was partially confirmed but ultimately negative.** Gradient clipping does stabilize T_max=50 and prevents the immediate divergence seen in #3045, but both gc values eventually hit the same gradient explosion — just delayed:

- **gc=1.0** delayed divergence from ep~60 (unclipped) to ep~52 — surprisingly *less* stable than unclipped in the mid-term, though it hit a better pre-divergence val (12.56% vs 10.41%). Diverged catastrophically with grad norms reaching Infinity by ep80.
- **gc=0.5** was significantly more stable, surviving until ep~121 and reaching val=7.28% — the best T_max=50 result by far. Each cosine cycle pushed the val lower (19% → 11% → 9.5% → 8.2% → 7.3%) before the eventual explosion.
- **Both eventually diverge.** The underlying instability of T_max=50's long cycles accumulates over many restarts. Tighter clipping (0.5) delays it much longer than loose clipping (1.0) but doesn't eliminate it.
- **Neither beats the baseline.** The champion at T_max=30 (val=3.997% @ ep467) benefits from 467 stable epochs of training. gc=0.5 + T_max=50 only got 120 stable epochs before diverging, and its best val (7.28%) is 1.82x worse than the baseline.

**Key insight:** T_max=50's longer cycles do find deeper per-cycle basins (gc=0.5 improved every cycle), but the gradient explosion is a fundamental property of the long cycle length, not just a clipping problem. T_max=30's inherent stability allows hundreds of epochs of stable training, which compounds into better final performance despite each individual cycle being shallower.

### Suggested follow-ups

1. **Adaptive/decaying gc**: Start with gc=0.5, reduce to gc=0.3 or gc=0.2 after epoch 100 to try to extend the stable window
2. **T_max=40 + gc=0.5**: A middle ground — longer cycles than T_max=30 but potentially more stable than T_max=50
3. **Early stopping at T_max=50**: Use gc=0.5 but stop at epoch 100 and evaluate — the model quality at ep107 (7.28%) suggests the trajectory was still improving when instability hit
4. **CosineAnnealingWarmRestarts with T_mult>1**: Increasing cycle lengths could combine the stability of short early cycles with the basin-finding of longer later cycles

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3114 because ADVISOR: Both T_max=50 variants diverge — gc=1.0: 12.56% diverged ep52; gc=0.5: 7.28% diverged ep121. The cycle-by-cycle improvement under gc=0.5 (19%→11%→9.5%→8.2%→7.3%) is interesting but the gradient explosion is intrinsic to the long-cycle LR peaks. T_max=30 provides sustained stability over 467+ epochs which T_max=50 cannot match. Closing — T_max=50 is conclusively dead for DM. T_max=30 confirmed as the only viable cosine period..
